### PR TITLE
feat: `signature_types` application arg.

### PR DESCRIPTION
**Pull request by peterschutt**
### Pull Request Checklist

- [x] New code has 100% test coverage
- [x] (If applicable) The prose documentation has been updated to reflect the changes introduced by this PR
- [x] (If applicable) The reference documentation has been updated to reflect the changes introduced by this PR
- [x] Pre-Commit Checks were ran and passed
- [x] Tests were ran and passed

### Description

- This PR adds a new argument to the app and layers, `signature_types`.

Types in this collection are added to `signature_namespace` using the type's `__name__` attribute.

The objective is to provide a nicer interface when adding names to the signature namespace without modifying the type name, e.g.: `signature_namespace={"Model": Model}` is equivalent to `signature_types=[Model]`.

The implementation makes it an error to supply a type in `signature_types` that has a value for `__name__` already in the signature namespace.

It will also throw an error if an item in `signature_types` has no `__name__` attribute.

### Close Issue(s)

-

### CODE REVIEW

**Comment by peterschutt** [comment]
Consider this one an RFC @litestar-org/maintainers - I'll build out tests if we choose to go ahead.

File: litestar/utils/signature.py [hunk 1]
```diff
@@ -157,3 +159,20 @@ def infer_request_encoding_from_field_definition(field_definition: FieldDefiniti
     if isinstance(field_definition.default, BodyKwarg):
         return field_definition.default.media_type
     return RequestEncodingType.JSON
+
+
+def add_types_to_signature_namespace(signature_types: list[Any], signature_namespace: dict[str, Any]) -> dict[str, Any]:
+    """Add types from a signature to a signature namespace.
```
Comment (peterschutt) [inline litestar/utils/signature.py:165]:
```suggestion
    """Add types from the list to a signature namespace.
```

File: docs/usage/routing/handlers.rst [hunk 2]
```diff
@@ -571,7 +571,7 @@ the module scope at runtime. We can address this on a case-by-case basis by sile
     # Choose the appropriate noqa directive according to your linter
     from domain import Model  # noqa: TCH002
 
-However, this approach can get tedious, so as an alternative, Litestar accepts a ``signature_namespace`` mapping at
+However, this approach can get tedious, so as an alternative, Litestar accepts a ``signature_types`` sequence at
```
Comment (provinzkraut) [inline docs/usage/routing/handlers.rst:574]: I would still mention the `signature_namespace`, maybe in an admontion, and explain the difference between the two.

**Review by provinzkraut — APPROVED** [review]
I like it!

File: litestar/app.py [hunk 3]
```diff
@@ -205,6 +205,7 @@ def __init__(
         return_dto: type[AbstractDTO] | None | EmptyType = Empty,
         security: Sequence[SecurityRequirement] | None = None,
         signature_namespace: Mapping[str, Any] | None = None,
+        signature_types: Sequence[Any] | None = None,
```
Comment (guacs) [inline litestar/app.py:208]: Couldn't the type be `Sequence[type[Any]]`?

File: litestar/config/app.py [hunk 4]
```diff
@@ -180,6 +180,11 @@ class AppConfig:
     """
     signature_namespace: dict[str, Any] = field(default_factory=dict)
     """A mapping of names to types for use in forward reference resolution during signature modelling."""
+    signature_types: list[Any] = field(default_factory=list)
```
Comment (guacs) [inline litestar/config/app.py:183]:
```suggestion
    signature_types: Sequence[Any] = field(default_factory=list)
```

File: litestar/controller.py [hunk 5]
```diff
@@ -144,6 +145,11 @@ class Controller:
     """A sequence of dictionaries that to the schema of all route handlers under the controller."""
     signature_namespace: dict[str, Any]
     """A mapping of names to types for use in forward reference resolution during signature modelling."""
+    signature_types: list[Any] | None
```
Comment (guacs) [inline litestar/controller.py:148]:
```suggestion
    signature_types: Sequence[Any]
```

Is there a reason to make it optional? It's not optional for `signature_namespace`.

**Review by guacs — APPROVED** [review]

File: litestar/config/app.py [hunk 6]
```diff
@@ -180,6 +180,11 @@ class AppConfig:
     """
     signature_namespace: dict[str, Any] = field(default_factory=dict)
     """A mapping of names to types for use in forward reference resolution during signature modelling."""
```
Comment (JacobCoffee) [inline litestar/config/app.py:182]:
```suggestion
    """A mapping of names to types for use in forward reference resolution during signature modeling."""
```

File: litestar/controller.py [hunk 7]
```diff
@@ -144,6 +145,11 @@ class Controller:
     """A sequence of dictionaries that to the schema of all route handlers under the controller."""
     signature_namespace: dict[str, Any]
     """A mapping of names to types for use in forward reference resolution during signature modelling."""
```
Comment (JacobCoffee) [inline litestar/controller.py:147]:
```suggestion
    """A mapping of names to types for use in forward reference resolution during signature modeling."""
```

File: litestar/controller.py [hunk 8]
```diff
@@ -144,6 +145,11 @@ class Controller:
...
+    signature_types: list[Any] | None
+    """A sequence of types for use in forward reference resolution during signature modelling.
```
Comment (JacobCoffee) [inline litestar/controller.py:149]:
```suggestion
    """A sequence of types for use in forward reference resolution during signature modeling.
```

File: litestar/handlers/base.py [hunk 9]
```diff
@@ -111,6 +112,8 @@ def __init__(
                 outbound response data.
             signature_namespace: A mapping of names to types for use in forward reference resolution during signature
                 modelling.
+            signature_types: A sequence of types for use in forward reference resolution during signature modelling.
```
Comment (JacobCoffee) [inline litestar/handlers/base.py:115]:
```suggestion
            signature_types: A sequence of types for use in forward reference resolution during signature modeling.
```

File: litestar/app.py [hunk 10]
```diff
@@ -291,6 +292,8 @@ def __init__(
                 See
                 :data:`SecurityRequirement <.openapi.spec.SecurityRequirement>` for details.
             signature_namespace: A mapping of names to types for use in forward reference resolution during signature modelling.
```
Comment (JacobCoffee) [inline litestar/app.py:294]:
```suggestion
            signature_namespace: A mapping of names to types for use in forward reference resolution during signature modeling.
```

File: litestar/app.py [hunk 11]
```diff
@@ -291,6 +292,8 @@ def __init__(
...
+            signature_types: A sequence of types for use in forward reference resolution during signature modelling.
```
Comment (JacobCoffee) [inline litestar/app.py:295]:
```suggestion
            signature_types: A sequence of types for use in forward reference resolution during signature modeling.
```

File: litestar/config/app.py [hunk 12]
```diff
@@ -180,6 +180,11 @@ class AppConfig:
...
+    signature_types: list[Any] = field(default_factory=list)
+    """A sequence of types for use in forward reference resolution during signature modelling.
```
Comment (JacobCoffee) [inline litestar/config/app.py:184]:
```suggestion
    """A sequence of types for use in forward reference resolution during signature modeling.
```

File: litestar/router.py [hunk 13]
```diff
@@ -148,6 +150,8 @@ def __init__(
                 See :data:`SecurityRequirement <.openapi.spec.SecurityRequirement>`
                 for details.
             signature_namespace: A mapping of names to types for use in forward reference resolution during signature modelling.
+            signature_types: A sequence of types for use in forward reference resolution during signature modelling.
+                These types will be added to the signature namespace using their ``__name__`` attribute.
             tags: A sequence of string tags that will be appended to the schema of all route handlers under the
```
Comment (JacobCoffee) [inline litestar/router.py:155]:
```suggestion
            signature_namespace: A mapping of names to types for use in forward reference resolution during signature modeling.
            signature_types: A sequence of types for use in forward reference resolution during signature modeling.
                These types will be added to the signature namespace using their ``__name__`` attribute.
            tags: A sequence of string tags that will be appended to the schema of all route handlers under the
```

File: litestar/testing/helpers.py [hunk 14]
```diff
@@ -215,6 +216,8 @@ def test_my_handler() -> None:
             See
             :data:`SecurityRequirement <.openapi.spec.SecurityRequirement>` for details.
         signature_namespace: A mapping of names to types for use in forward reference resolution during signature modelling.
+        signature_types: A sequence of types for use in forward reference resolution during signature modelling.
+            These types will be added to the signature namespace using their ``__name__`` attribute.
```
Comment (JacobCoffee) [inline litestar/testing/helpers.py:220]:
```suggestion
        signature_namespace: A mapping of names to types for use in forward reference resolution during signature modeling.
        signature_types: A sequence of types for use in forward reference resolution during signature modeling.
            These types will be added to the signature namespace using their ``__name__`` attribute.
```

File: litestar/testing/helpers.py [hunk 15]
```diff
@@ -464,6 +469,8 @@ def test_my_handler() -> None:
             See
             :data:`SecurityRequirement <.openapi.spec.SecurityRequirement>` for details.
         signature_namespace: A mapping of names to types for use in forward reference resolution during signature modelling.
+        signature_types: A sequence of types for use in forward reference resolution during signature modelling.
```
Comment (JacobCoffee) [inline litestar/testing/helpers.py:472]:
```suggestion
        signature_namespace: A mapping of names to types for use in forward reference resolution during signature modeling.
        signature_types: A sequence of types for use in forward reference resolution during signature modeling.
```

**Review by JacobCoffee — APPROVED** [review]

File: litestar/app.py (hunk 3 above)
Comment (peterschutt) [reply litestar/app.py:208]: Not all valid type-annotations are types though, and this is consistent with how we have typed `signature_namespace`.

```py
>>> from inspect import isclass
>>> from typing import Union
>>> isclass(Union)
False
```

File: litestar/config/app.py (hunk 4 above)
Comment (peterschutt) [reply litestar/config/app.py:183]: As we control the instantiation of the config object:

https://github.com/litestar-org/litestar/blob/b6fc70cf2cda7c42a53bc224622d45d1fd4a304c/litestar/app.py#L319-L368

...then, we know these are concrete collection types, not abstract types.

The advantage of this is for plugins. If this is typed `list[Any]`, a plugin can readily append to it:

```py
config.signature_types.append(PluginManagedType)
```

If typed as `Sequence`:

```py
if not isinstance(config.signature_types, list):
    config.signature_types = list(config.signature_types)
config.signature_types.append(PluginManagedType)
```

File: litestar/controller.py (hunk 5 above)
Comment (peterschutt) [reply litestar/controller.py:148]: I don't think there is a reason for optional here. And `Sequence` should be appropriate too. Also, `signature_namespace` should be `Mapping` as well. Cheers!

Comment (peterschutt) [reply litestar/controller.py:148]: I resisted the urge to change the `signature_namespace` for now, mostly just to keep this PR focused.

File: litestar/config/app.py (hunk 4 above)
Comment (guacs) [reply litestar/config/app.py:183]: Aah cool that makes sense.

**Pull request merged**